bitmap/bitmap.go

```diff
@@ -0,0 +1,105 @@
+package bitmap
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/RoaringBitmap/roaring"
+)
+
+type Bitmap struct {
+	base  int32
+	size  int32
+	start int32
+	data  *roaring.Bitmap
+	lock  *sync.Mutex
+}
+
+// NewBitmap allocate a bitmap range from [start, end], notice the end is included
+func NewBitmap(start, end int32) *Bitmap {
+	size := end - start + 1
+	data := roaring.New()
+	if size > 0 {
+		data.AddRange(0, uint64(size))
+	}
+	return &Bitmap{
+		base:  start,
+		size:  size,
+		start: int32(0),
+		data:  data,
+		lock:  &sync.Mutex{},
+	}
+}
+
+func (b *Bitmap) findAt(count, iStart int32) (int32, int32, error) {
+	i := b.data.Iterator()
+	i.AdvanceIfNeeded(uint32(iStart))
+	bStart := iStart
+	for bStart <= b.size {
+		last := int32(-1)
+		remains := count
+		for i.HasNext() && remains > 0 {
+			// first element
+			if last < 0 {
+				last = int32(i.Next())
+				bStart = last
+				remains--
+				continue
+			}
+			next := int32(i.Next())
+			// failed to find the available range
+			if next-last > 1 {
+				break
+			}
+			last = next
+			remains--
+		}
+		if remains == 0 {
+			break
+		}
+		if !i.HasNext() {
+			return 0, 0, fmt.Errorf("cannot find an empty port range")
```

Comment:
I just realized that the partially allocated ports above cannot be released if the function errors out here.
This issue exists in the previous implementation as well.